fix(replays): display unique frame colors in prioritized order

michellewzhang opened the pull request:
Fixes #50308 by displaying the frame colors in the replays timeline in a prioritized order, and only displaying unique colors (e.g. not displaying 2 rings of green, only displaying 1 ring of green).

Before prioritization:
<img width="569" alt="Screenshot 2023-07-13 at 1 49 51 PM" src="https://github.com/getsentry/sentry/assets/56095982/b65e7b7c-5eca-47ef-8626-a261cf8e345b">
Colors are not sorted by priority, leading the red color to be hidden.

After prioritization:
<img width="352" alt="Screenshot 2023-07-13 at 1 49 31 PM" src="https://github.com/getsentry/sentry/assets/56095982/34f98f32-4a78-4c6b-ba9d-e5b9d712a18c">
Red is now the outermost ring.

Before only displaying unique colors:
<img width="541" alt="Screenshot 2023-07-13 at 1 50 02 PM" src="https://github.com/getsentry/sentry/assets/56095982/a940c606-b451-406f-9a62-a58485930167">
Bubble size is larger.

After only displaying unique colors:
<img width="386" alt="Screenshot 2023-07-13 at 1 49 45 PM" src="https://github.com/getsentry/sentry/assets/56095982/d5b5acb7-8446-4b3b-8edb-b710daada731">
Bubble size is smaller because only unique colors are displayed.

Zylphrex committed:
feat(functions): Split function trends stats query (#52588)

Querying 50 timeseries can be slow. This splits the query into batches
of 10 so it can be parallelized.

RyanSkonnord committed:
fix(hc): Tolerate OrganizationMapping not found when resolving region (#52560)

Introduce the `return_none_if_mapping_not_found` option to the
`regional_rpc_method`, which causes the decorated method to return
`None` indicating the queried object does not exist if we cannot find a
region in which to look for it in the first place.

markstory committed:
fix(hybridcloud) Fix typing & loop error in silo router (#52753)

Fix a typing issue and condition within a loop that should be outside of
the loop.

malwilley committed:
feat(issue-details): Add analytics for clicking open replay from issue details (#52762)

leeandher committed:
feat(hybrid-cloud): Add request parser for Jira Server (#52687)

This PR adds a request parser for incoming traffic from Jira Server
integrations and some associated tests.

schew2381 committed:
nit: Organize permission scopes on user tokens page (#52759)

Tried to figure out how to put it in cols instead of rows but was taking too long so this is a good incremental improvement

Before:
![Screenshot 2023-07-12 at 2 44 19
PM](https://github.com/getsentry/sentry/assets/67301797/cfefe9d8-bfe4-4f74-a12c-89a77a3545db)

After:
![Screenshot 2023-07-12 at 2 43 53
PM](https://github.com/getsentry/sentry/assets/67301797/e0b076b0-4246-47fc-afcb-662c2dbaae47)

dashed committed:
chore(hybrid-cloud): Update get_users_with_only_one_integration_for_provider to use OrganizationMemberMapping (#52763)

dashed committed:
chore(hybrid-cloud): Remove Integration model dependency in IntegrationRepositoryProvider. (#52691)

obostjancic committed:
feat(alerts): common field support for on-demand metrics (#52612)

priscilawebdev committed:
ref(getting-started-docs): Migrate React-Native doc to sentry main repo (#52729)

priscilawebdev committed:
fix(getting-started): Fix wrong text in the react native doc (#52782)

philipphofmann committed:
chore(sdk-crashes): Info on custom stacktrace rule (#52781)

Add a comment on custom stack trace rule.

philipphofmann committed:
docs(sdk-crashes): Add readme (#52783)

Add a short readme for the SDK crash detection.

obostjancic committed:
feat(alerts): common field support for on-demand metrics (#52613)

antonpirker committed:
Fix: Errors in Crons show up in wrong monitor. (#52788)

This will fix a problem with Celery Beat crons all having the same
trace. Now every Celery task started by Celery Beat will have its own
trace (preventing errors from one task showing up in another task)

Zylphrex committed:
styles(profiling): Adjust some minor styling on profiling widgets (#52755)

Some minor adjustments to clean up some of the styling on the profiling
widgets for alignment, clarity, etc.

asottile-sentry committed:
ref: upgrade django to 3.2 (#52713)

same as #52269 -- but I ran out of time to deploy it

jan-auer committed:
feat(on-demand-metrics): Support count_if in rule conversion (#52784)

Parses `count_if` fields and converts them to a `RuleCondition`. The
condition is then merged with conditions from the query. Also adds
support for `count_if` to `is_on_demand_query`.

Closes #52504

priscilawebdev committed:
ref(getting-started-docs): Migrate spring-boot  doc to sentry main repo (#52786)

EricHasegawa committed:
ref(auth): Add security as a codeowner for auth files (#52761)

We have a bunch of sensitive auth files living in our top level
`web/frontend/` folder. This adds the security team as an owner for the
relevant files. [Addresses comments in this
PR](https://github.com/getsentry/sentry/pull/52556).

Note that I plan to reorganize our auth files in the future so we can
move sensitive code into just a few main folders (and we won't have to
specify ownership on a file-by-file basis.

priscilawebdev committed:
ref(getting-started-docs): Migrate laravel doc to sentry main repo (#52789)

priscilawebdev committed:
ref(getting-started-docs): Remove no longer used hook & component (#52791)

udameli committed:
feat(trends): Handle query timeouts (#52745)

- reduce the top event count from 50 to 45 and increase transaction name
count per query from 10 to 15 until new metric is materialized. this is
a load improvement meant to reduce the number of concurrent queries from
5 to 3.
- update limits in case there were any api-level rate limits 
- prune unused project slugs to reduce cardinality

cmanallen committed:
feat(replays): Create click rows on dead and rage click events (#52456)

Pushes click events to snuba when a "dead" or "rage" click breadcrumb is
found. This pull also removes publishing to the issues platform as dead
and rage clicks were determined to not have enough utility to create
issues for.

markstory committed:
chore(hybridcloud) Remove role switching and update guard method (#52673)

The role switching logic to catch write operations that are missing
outboxes has been causing test flakiness and disrupting development
workflows.

These changes finish replacing role switching with query auditing that
is done during test teardown. Switching to a teardown based audit means
we can't give the precise location of where the bad operation was as the
stack frames have been lost. However, we are able to show the offending
query and its location in the SQL logs for a test.

getsentry-bot committed:
Revert "chore(hybridcloud) Remove role switching and update guard method (#52673)"

This reverts commit 07c498ab53b528ffc364bd1fa682e67207facf37.

Co-authored-by: markstory <24086+markstory@users.noreply.github.com>

edwardgou-sentry committed:
feat(starfish): Removes tables automatically sorting when datazooming on charts in starfish (#52741)

We used to sort tables automatically when zooming in on charts in span
summary and the module views. This removes that functionality since it
was confusing.

Abdkhan14 committed:
feat(perf-detector-threshold-configuration) Fixed dark mode styling a… (#52800)

- Dark mode styling off:
<img width="1038" alt="Screenshot 2023-07-13 at 11 02 29 AM"
src="https://github.com/getsentry/sentry/assets/60121741/ebc2419e-1085-4d33-92d2-6cfe619e0c32">

- Tooltip position off, needs to be closer to the slider track:
<img width="954" alt="Screenshot 2023-07-13 at 11 04 32 AM"
src="https://github.com/getsentry/sentry/assets/60121741/7ebd314f-6ac8-4e3d-8dd9-532363073e6e">

Co-authored-by: Abdullah Khan <abdullahkhan@PG9Y57YDXQ.local>

0Calories committed:
fix(starfish): None project if pagefilters not ready (#52760)

azaslavsky committed:
ref(backup): Use a more exact datetime serializer (#52771)

The default DjangoJSONEncoder inherits from the odd behavior of Python's
own `isoformat` serializer: it includes milliseconds in the output in
all cases EXCEPT those where the value of the milliseconds are `.000`,
in which case it removes them completely. This makes it difficult and
flaky to do simple string comparison between two dates. Instead, a new
wrapper class is introduced to ensure that we always retain the
milliseconds on all date serializations.

EricHasegawa committed:
feat(auth): Return basic user info in OIDC ID token (#52556)

Returns basic user info inside the ID token we issue from the `/token`
endpoint

markstory committed:
chore(hybridcloud) Move unguarded_write out of testutils (#52807)

When opening the mulligan pull requests, I remembered that I'll also
need to move the unguarded_write function out of testutils first, as I
need to update getsentry before sentry.

asottile-sentry committed:
ref: fix some misc mypy errors (#52765)

<!-- Describe your PR here. -->

cathteng committed:
fix(github-comment): add github comment task referrer (#52809)

armenzg committed:
ref(save_event): Refactor a bit (#52792)

Changes:
* Rename a span for secondary grouping
* Only auto upgrade group config when saving an error
* Move logic to save error into its own function

---------

Co-authored-by: Armen Zambrano G <armenzg@users.noreply.github.com>

spalmurray committed:
ref(stats): Convert teamReleases to functional (#52413)

Also added some test cases for teamReleases

spalmurray committed:
feat(discord): Implement Discord interactions endpoint (#52498)

Implement minimal interactions endpoint that is able to receive,
validate, and respond to PING events from Discord.

barkbarkimashark committed:
fix(helpful-event): sort on whether replay_id or profile_id are null or not and not the actual id values (#52777)

Fixes an issue where we sort on the ID values themselves instead of
sorting on whether the replay_id or profile_id are present.

barkbarkimashark committed:
feat(group-attributes): log a metric when certain fields for Group and related tables are updated (#52646)

We're gathering metrics when any of the following postgres DB tables are
mutated:
- Group, GroupAssignee, GroupOwner rows are created or deleted
- Group.status, Group.substatus, Group.first_seen, Group.num_comments
are updated

These metrics will help us gauge how 'active' these columns are for the
purposes of replicating these values to snuba.

billyvg committed:
chore(sdk): Upgrade SDKs to 7.59.0-beta.0 (#52805)

See
https://github.com/getsentry/sentry-javascript/releases/tag/7.59.0-beta.0

feat(replay): Re-sample sessions after they expire #8407
This pre-release contains a revamped replay session expiration handling.
Now, when a session expires (because it ran for too long), we will
always
re-sample the session, and may start another one (or start buffering)
based on the sample rates.

rjo100 committed:
feat(crons): Add helpful subtitle to crons issue platform + update evidence (#51877)

Adds helpful human-readable subtitle to issue platform crons issues +
updates `last check-in` evidence to `last successful check-in`

Zylphrex committed:
feat(functions): Pass historical data for function trends (#52801)

To help the breakpoint detection avoid false positive, we want to use
the first 20% of the data as historicial data. This allows the detector
to learn any periodicity in the data and decreases the likelihood of
showing false positives in these cases.

priscilawebdev committed:
fix(getting-started-docs): Fix code snippets indentation (#52804)

Commit:
Remove replay search alert banner

snigdhas committed:
feat(escalating-issues): Add new use case for metrics dataset (#52360)

Register a new use case for the metrics dataset to be used with
escalating issues.

edwardgou-sentry committed:
feat(starfish): Fixes spans view operation, action, and domain filter dropdown items to match actual results (#52659)

Dropdown options for operation, action, and domain filter were not
matching the actual unique values in the table result. This is because
the query filters for the operation, action, and domain requests didnt
match the filter for the table request.

asottile-sentry committed:
ref: fix typing for integration_proxy (#52810)

<!-- Describe your PR here. -->

azaslavsky committed:
test(backup): Add monitor tests (#52485)

Tests for monitoring.

Issue: getsentry/team-ospo#156

getsentry-bot committed:
Revert "feat(crons): Add helpful subtitle to crons issue platform + update evidence (#51877)"

This reverts commit b466fbcefb0e2df7fa9452e7f6383bddc2953def.

Co-authored-by: rjo100 <7078270+rjo100@users.noreply.github.com>

0Calories committed:
fix(starfish): Filter out empty actions from dropdown (#52757)

vuluongj20 committed:
ref(crons): Use darker colors for timeline marks (#52818)

**Before ——**
<img width="1589" alt="image"
src="https://github.com/getsentry/sentry/assets/44172267/080fa20c-2f4f-4d36-b903-eebf1210e4d5">


**After ——**
<img width="1589" alt="image"
src="https://github.com/getsentry/sentry/assets/44172267/f61affd6-a2cd-4ff5-af89-eeb8b4053b93">

DominikB2014 committed:
fix(starfish): span sample back button behaviour (#52806)

Fixes a bug where clicking the browser back button when the span samples
page is open leads to a `undefined` span samples page.

markstory committed:
chore(hybridcloud) Remove role switching and update guard method v2  (#52802)

Mulligan on https://github.com/getsentry/sentry/pull/52673 as using
testutils code in application code blows up docker image generation.

Because of that I've moved unguarded_write to sentry.silo instead.

Zylphrex committed:
fix(functions): Switch to fingerprints when querying functions (#52815)

It's more effectively to query for a function by it's fingerprint than
to query by it's name.

ceorourke committed:
ref(alerts): Remove mute alerts feature flag (#52822)

Remove usages of the mute metric alerts feature flag now that it's in
GA.

Back end PR: https://github.com/getsentry/sentry/pull/52823

asottile-sentry committed:
ref: remove unnecessary test from sudo.test_utils (#52808)

this was just re-testing django functions, not ours




<!-- Describe your PR here. -->

leeandher committed:
feat(hybrid-cloud): Create parser for Github Enterprise (#52813)

This PR adds the Github Enterprise request parser. This parser is nearly
identical to the regular Github one, so it actually inherits it and
makes a few changes to the 'external_id' code. I also modified the code
at the webhook application code so they would refer to the same method
of getting an integration and we wouldn't introduce any drift by
accident.

---------

Co-authored-by: Mark Story <mark@mark-story.com>

rjo100 committed:
feat(crons): Add helpful subtitle to crons issue platform + update evidence (#52817)

Adds helpful human-readable subtitle to issue platform crons issues +
updates last check-in evidence to last successful check-in

malwilley committed:
fix(helpful-event): Prevent event from being refetched when using All Events tab (#52752)

Fixes https://github.com/getsentry/sentry/issues/52738
Fixes SENTRY-131D

`useApiQuery` refetches whenever its query key changes. Since the
`query` part of the URL can change when you search in the "All Events"
tab, this was causing refetches of the helpful event. This does two
things to prevent that from occurring:

1. Disables the event details query when we are not on the details tab
2. Clears `query` from the query string when navigating away from the
All Events tab

mrduncan committed:
fix(hc): Move OrganizationApiKeyDetailsEndpoint to control silo (#52776)

The ApiKey model lives in the control silo.

IanWoodard committed:
fix(replays): Set max-width of progress to 100% (#52816)

Fixes #52304 by setting the max width of the progress to 100%.

I do find it odd that the context allows the `currentTime` to be be
larger than `durationMs`, but this fix ensures that the progress will
never appear wider than 100% (which I think is good to have regardless).

nhsiehgit committed:
feat(Fly Partnership): Partnership auth route (#52772)

Set up a new auth route specifically for partners so they can direct
users to Sentry with their organization_id (NOT sentry organization_id).

We construct a config from the passed query params and fetch the
AuthProvider model instance to pull the org id from

0Calories committed:
fix(starfish): Fix WSV bug where Other module is added when it does not exist (#52831)

scttcper committed:
feat(search-shortcuts): Group assignee search suggestions (#52764)

markstory committed:
chore: Remove role related methods (#52826)

Now that they are unused we can remove the role switching logic.

Commit:
fix(replays): display unique frame colors in prioritized order

cathteng committed:
feat(github-growth): allow hiding repos (#52766)

mrduncan committed:
fix(hc): Move OrganizationApiKeyIndexEndpoint to control silo (#52824)

The ApiKey model lives in the control silo.

#52776 took care of the details endpoint

cathteng committed:
feat(github-growth): exclude hidden repos (#52773)

ryan953 commented on static/app/components/replays/breadcrumbs/replayTimelineEvents.tsx:
```diff
@@ -97,16 +97,26 @@ function Event({
   `;
 
-  // If we have more than 3 events we want to make sure of showing all the different colors that we have
+  // We want to make sure of showing all the different colors that we have
```
```suggestion
  // We want to show the full variety of colors available.
```

ryan953 commented on static/app/components/replays/breadcrumbs/replayTimelineEvents.tsx:
```diff
@@ -97,16 +97,26 @@ function Event({
+
+  // Sort the frame colors by color priority
+  // Priority order: red300, yellow300, green300, purple300, gray300
+  const sortedUniqueColors: Color[] = uniqueColors.sort(function (x, y) {
+    const colorOrder = ['red300', 'yellow300', 'green300', 'purple300', 'gray300'];
```
```suggestion
  const sortedUniqueColors = uniqueColors.sort(function (x, y) {
    const colorOrder: Color[] = ['red300', 'yellow300', 'green300', 'purple300', 'gray300'];
```

ryan953 commented on static/app/views/replays/list/replaysList.tsx:
seems like some extra stuff got added into the PR

barkbarkimashark committed:
fix(helpful-event): properly pass in a sequence of environment.names instead of sequence of Environment objects (#52839)

`get_latest_event_for_environments()` function takes in a sequence of
environment names and not sequence of `Environment` objects.

Resolves SENTRY-1378

EricHasegawa commented:
the sentry rite of passage
